src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -29,75 +33,115 @@ NaiveKMeans<DistanceType, MatType>::NaiveKMeans(const MatType& dataset,
+
+    arma::mat& localCentroids = threadCentroids[threadId];
+    arma::Col<size_t>& localCounts = threadCounts[threadId];
+
+    arma::vec distances(clusters);
+
+    for (size_t i = segmentStart; i < segmentEnd; ++i)
     {
-      // Find the closest centroid to this point.
-      double minDistance = std::numeric_limits<double>::infinity();
-      size_t closestCluster = centroids.n_cols; // Invalid value.
+      const auto dataPoint = dataset.col(i);
+      const double dataNorm = arma::dot(dataPoint, dataPoint);
 
-      for (size_t j = 0; j < centroids.n_cols; ++j)
+      // Calculate distances to all centroids
+      for (size_t j = 0; j < clusters; ++j)
       {
-        const double dist = distance.Evaluate(dataset.col(i),
-            centroids.unsafe_col(j));
-        if (dist < minDistance)
-        {
-          minDistance = dist;
-          closestCluster = j;
-        }
+        const arma::vec& centroid = centroids.col(j);
+        distances(j) = std::max(0.0, dataNorm + centroidNorms(j) - 2 * arma::dot(dataPoint, centroid));
       }
 
-      Log::Assert(closestCluster != centroids.n_cols);
+      // Find the closest centroid
+      const size_t closestCluster = distances.index_min();
 
-      // We now have the minimum distance centroid index.  Update that centroid.
-      localCentroids.unsafe_col(closestCluster) += dataset.col(i);
+      // Update local centroids and counts
+      localCentroids.col(closestCluster) += dataPoint;
       localCounts(closestCluster)++;
     }
-    // Combine calculated state from each thread
-    #pragma omp critical
-    {
-      newCentroids += localCentroids;
-      counts += localCounts;
-    }
   }
 
-  // Now normalize the centroid.
-  for (size_t i = 0; i < centroids.n_cols; ++i)
-    if (counts(i) != 0)
-      newCentroids.col(i) /= counts(i);
+  // Combine results from all threads
+  for (size_t t = 0; t < effectiveThreads; ++t)
+  {
+    newCentroids += threadCentroids[t];
+    counts += threadCounts[t];
+  }
 
-  distanceCalculations += centroids.n_cols * dataset.n_cols;
+  const double eps = std::numeric_limits<double>::epsilon();
+  // Normalize the centroids
+  for (size_t j = 0; j < clusters; ++j)
+  {
+    if (counts(j) > eps)
+    {
+      newCentroids.col(j) /= counts(j);
+    }
+    else
+    {
+      // Handle empty or near-empty cluster
+      newCentroids.col(j) = centroids.col(j);
+    }
+  }
 
-  // Calculate cluster distortion for this iteration.
+  // Calculate cluster distortion
   double cNorm = 0.0;
-  for (size_t i = 0; i < centroids.n_cols; ++i)
+  #ifdef MLPACK_USE_OPENMP
+    #pragma omp parallel for reduction(+:cNorm) schedule(static)
+  #endif
+  for (size_t j = 0; j < clusters; ++j)
   {
-    cNorm += std::pow(distance.Evaluate(centroids.col(i), newCentroids.col(i)),
-        2.0);
+    cNorm += arma::norm(centroids.col(j) - newCentroids.col(j), 2);
```

Comment:
This actually makes this computation incorrect if `distance` is not `EuclideanDistance`.